### PR TITLE
update to hinge loss

### PR DESCRIPTION
**Pull request by vict0rsch**
(/!\ PR based off of the `infer` branch, not master. Should not impact this particular bit but still, be aware)

@51N84D (and others) I need you to be careful in reviewing this PR:

* there's a new parameter called `for_discriminator` which I set to False in the gen's update and True in the disc's update
* you have a global and a local loss. Not sure the vanilla way I updated to hinge loss makes the most sense so I need your input on this
  * especially, SPADE's hinge loss has an internal way of dealing with multi resolution so I think we can get rid of the for loop _around_ the loss. If you agree I'll let you update

critical thinking's required here otherwise it'll be trained to nonsense ;)

### CODE REVIEW

File: omnigan/losses.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ def get_losses(opts, verbose, device=None):
     # ------------------------------
     # painter losses
     if "p" in opts.tasks:
-        losses["G"]["p"]["gan"] = GANLoss()
+        losses["G"]["p"]["hinge"] = HingeLoss()
```
Comment (vict0rsch) [inline omnigan/losses.py:370]: note: this has changed name

File: omnigan/losses.py [hunk 2]
```diff
@@ -413,9 +413,7 @@ def get_losses(opts, verbose, device=None):
     # ----------------------------------
     # -----  Discriminator Losses  -----
     # ----------------------------------
-    losses["D"]["default"] = GANLoss(
-        soft_shift=opts.dis.soft_shift, flip_prob=opts.dis.flip_prob, verbose=verbose
-    )
+    losses["D"]["p"] = HingeLoss()
```
Comment (vict0rsch) [inline omnigan/losses.py:416]: default did not make sense, it's the painter disc

File: omnigan/losses.py [hunk 3]
```diff
@@ -478,3 +476,50 @@ def multiDiscriminatorAdapter(d_out, opts):
             "Check the setting of OmniDiscriminator! For now, we don't support multi-scale Omnidiscriminator."
         )
     return d_out
+
+
+class HingeLoss(nn.Module):
+    """
+    Adapted from https://github.com/NVlabs/SPADE/blob/master/models/networks/loss.py
+    for  the painter
```
Comment (vict0rsch) [inline omnigan/losses.py:484]: note: I took it from a repo so I add the source ;)

File: omnigan/trainer.py [hunk 4]
```diff
@@ -236,6 +236,11 @@ def setup(self, inference=False):
         )
 
         if self.input_shape is None:
+            if inference:
```
Comment (vict0rsch) [inline omnigan/trainer.py:239]: not relevant, just an artifact  from my concurrent work on `infer`

File: omnigan/trainer.py [hunk 5]
```diff
@@ -944,8 +949,12 @@ def get_painter_loss(self, multi_domain_batch):
                 # Take last element for GAN loss on discrim prediction
                 update_loss = (
                     (
-                        self.losses["G"]["p"]["gan"](fake_d_global[i][-1], True)
-                        + self.losses["G"]["p"]["gan"](fake_d_local[i][-1], True)
+                        self.losses["G"]["p"]["hinge"](
+                            fake_d_global[i][-1], True, False
```
Comment (vict0rsch) [inline omnigan/trainer.py:953]: new arg: `for_discriminator`

File: omnigan/trainer.py [hunk 6]
```diff
@@ -944,8 +949,12 @@ def get_painter_loss(self, multi_domain_batch):
...
+                        )
+                        + self.losses["G"]["p"]["hinge"](
+                            fake_d_local[i][-1], True, False
+                        )
                     )
                     * lambdas.G["p"]["gan"]
                     / num_D
```
Comment (vict0rsch) [inline omnigan/trainer.py:960]: kept the `gan` namespace not to have to change configs

File: omnigan/trainer.py [hunk 7]
```diff
@@ -1090,13 +1099,13 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
                 for i in range(num_D):
```
Comment (vict0rsch) [inline omnigan/trainer.py:1099]: Maybe this loop is equivalently dealt with within the HingeLoss

File: omnigan/losses.py (hunk 2 above)
Comment (51N84D) [reply omnigan/losses.py:416]: Hmm is `losses["D"]["default"]` not used outside of the painter? I thought it makes sense to have some standard discriminator loss

Comment (vict0rsch) [reply omnigan/losses.py:416]: I agree, but it's not

**Review by alexrey88 — APPROVED** [review]
